File: Problem_22_Generate_Parentheses/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        backtrack(result, "", 0, 0, n);
        return result;
    }

private:
    void backtrack(vector<string>& result, string current, int openCount, int closeCount, int n) {
        if (current.size() == 2 * n) {
            result.push_back(current);
            return;
        }

        if (openCount < n) {
            backtrack(result, current + '(', openCount + 1, closeCount, n);
        }
        if (closeCount < openCount) {
            backtrack(result, current + ')', openCount, closeCount + 1, n);
        }
    }
};

File: Problem_22_Generate_Parentheses/solution.py
from typing import List


def generate_parentheses(n: int) -> List[str]:
    result = []

    def backtrack(current: str, open_count: int, close_count: int) -> None:
        if len(current) == 2 * n:
            result.append(current)
            return

        if open_count < n:
            backtrack(current + '(', open_count + 1, close_count)
        if close_count < open_count:
            backtrack(current + ')', open_count, close_count + 1)

    backtrack('', 0, 0)
    return result


if __name__ == "__main__":
    print(generate_parentheses(3))
